examples/cis2-wccd-upgradable/src/lib.rs

```diff
@@ -0,0 +1,3181 @@
+//! on the `implementation` before invoking these functions. Invoking the wccd
+//! functions directly on the `implementation` without going through the
+//! `proxy` fallback function will revert.
+
+//! The admin address on the `proxy` can upgrade the protocol with
+//! a new `implementation` contract. The admin address on the
+//! `implementation` can pause/unpause the protocol.
+//!
+//! Deploy the `state` and the `implementation` contract first by invoking their
+//! respective `init` functions. Then, deploy the `proxy` contract and pass
+//! the already deployed contract addresses into the proxy `init` function.
+//! Then, call the `initialize` functions on the `state` as well as on the
+//! `implementation` contracts to set the remaining addresses.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use concordium_cis2::*;
+use concordium_std::{fmt::Debug, *};
+
+/// The id of the wCCD token in this contract.
+const TOKEN_ID_WCCD: ContractTokenId = TokenIdUnit();
+
+/// The metadata url for the wCCD token.
+const TOKEN_METADATA_URL: &str = "https://some.example/token/wccd";
+
+// Types
+
+/// Contract token ID type.
+/// Since this contract will only ever contain this one token type, we use the
+/// smallest possible token ID.
+type ContractTokenId = TokenIdUnit;
+
+/// Contract token amount type.
+/// Since this contract is wrapping the CCD and the CCD can be represented as a
+/// u64, we can specialize the token amount to u64 reducing module size and cost
+/// of arithmetics.
+type ContractTokenAmount = TokenAmountU64;
+
+/// Needed for the custom serial instance, which doesn't include the `Option`
+/// tag and the length of the vector.
+#[derive(PartialEq, Eq, Debug)]
+struct RawReturnValue(Option<Vec<u8>>);
+
+impl Serial for RawReturnValue {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        match &self.0 {
+            Some(rv) => out.write_all(rv),
+            None => Ok(()),
+        }
+    }
+}
+
+/// The state tracked for each address.
+#[derive(Serial, DeserialWithState, Deletable)]
+#[concordium(state_parameter = "S")]
+struct AddressState<S> {
+    /// The number of tokens owned by this address.
+    balance:   ContractTokenAmount,
+    /// The address which are currently enabled as operators for this token and
+    /// this address.
+    operators: StateSet<Address, S>,
+}
+
+/// The `state` contract state.
+#[derive(Serial, DeserialWithState)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    /// Address of the w_ccd proxy contract.
+    proxy_address:          Option<ContractAddress>,
+    /// Address of the w_ccd implementation contract.
+    implementation_address: Option<ContractAddress>,
+    /// The state of the one token.
+    token:                  StateMap<Address, AddressState<S>, S>,
+    /// Timestamp when contract is unpaused.
+    unpause_time:           Timestamp,
+}
+
+/// The `implementation` contract state.
+#[derive(Serial, Deserial, SchemaType)]
```

Comment:
Then I guess it depends on the purpose of those `view` functions. When will they be used and by who? Perhaps it makes more sense to create new types used for purpose of viewing.

Reply:
The view functions are there for manual checking of the state of the contracts. E.g. you might want to check if after the setup all contracts are correctly linked up. The function will be used mainly by us but also by some interested users who want to explore the protocol. 

Not sure if we want to add even more types. There are already a lot of different types in the contract and since the `StateProxy` and `StateImplementation` types are working fine as a return parameter for the `view` function, I am not so convinced that we need a new type here.

Reply:
Ok, then I think it's fine to keep it as is.